### PR TITLE
feat(tts,stt): add StreamAdapter

### PR DESCRIPTION
**Pull request by nbsp**

### CODE REVIEW

**Comment by nbsp** [comment]
STT seems to work well enough, but the TTS StreamAdapter fails in weird ways.

**Comment by nbsp** [comment]
only thing left to tackle: OpenAI TTS doesn't allow for function calling. i assume the culprit is speech never ending on function calls, because there's no text to be called

**Review by theomonnom — APPROVED** [review]

File: agents/src/tts/stream_adapter.ts [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+    const synthesize = async () => {
+      for await (const ev of this.#sentenceStream) {
+        for await (const audio of this.#tts.synthesize(ev.token)) {
+          this.queue.put(audio);
+        }
+      }
+      this.queue.put(SynthesizeStream.END_OF_STREAM);
+    };
+
+    Promise.all([forwardInput(), synthesize()]);
```
Comment (theomonnom) [inline agents/src/tts/stream_adapter.ts:61]: How is this closed?

Comment (nbsp) [reply agents/src/tts/stream_adapter.ts:61]: forwardInput gets closed via endInput to the stream, which flushes the queue and eventually synthesize finishes processing

**Comment by jayeshp19** [comment]
Preparing review...

**Comment by jayeshp19** [comment]
## PR Reviewer Guide 🔍

<table>
<tr><td>⏱️&nbsp;<strong>Estimated effort to review</strong>: 4 🔵🔵🔵🔵⚪</td></tr>
<tr><td>🧪&nbsp;<strong>No relevant tests</strong></td></tr>
<tr><td>🔒&nbsp;<strong>No security concerns identified</strong></td></tr>
<tr><td>⚡&nbsp;<strong>Key issues to review</strong><br><br>

<a href='https://github.com/livekit/agents-js/pull/156/files#diff-312b25afdfc7dd2bd0c04a4f247005830b12913f3cab67606e4c33d3b5ef2f84R137-R139'><strong>Code Duplication</strong></a><br>The logic for checking if the stream should be terminated due to cancellation or end-of-stream is duplicated across multiple files (`agent_output.ts`, `agent_playout.ts`). Consider refactoring this into a shared function or utility to reduce duplication and improve maintainability.

<a href='https://github.com/livekit/agents-js/pull/156/files#diff-54d355ae58916188576582555b5551e4d7ee8302d1d5df85d8181038f623d2dbR73-R74'><strong>Error Handling</strong></a><br>The method `#run` in `StreamAdapterWrapper` class uses `Promise.all` to run `forwardInput` and `recognize` concurrently without proper error handling. If one of these promises rejects, the other will continue executing, which might lead to unhandled promise rejections or inconsistent states. Consider adding error handling mechanisms to manage such cases effectively.

<a href='https://github.com/livekit/agents-js/pull/156/files#diff-02d2a7363c069daee7507de8fb62dbb5ba76ba7120b18c39ce8ccd638d262783R61-R62'><strong>Resource Management</strong></a><br>The method `#run` in `StreamAdapterWrapper` class does not handle the scenario where the stream might need to be closed or cleaned up if an error occurs or when it ends. This can lead to resource leaks. It's important to ensure that resources are properly cleaned up in error scenarios or when the stream is no longer needed.

</td></tr>
</table>

**Pull request merged**